Feature/35716 changing owner flow

burczu opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR introduces ability to change ownership of the paid workspace as part of the Simplified collect project.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/35716
PROPOSAL: n/a


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Log in to the App as an admin of the paid workspace **who is not an owner** at the same time
2. Open paid workspace
3. Go to Members page
4. Click on the owner of the workspace
5. Click "Transfer owner"
6. Follow instructions and check if everything works correctly

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Changing ownership is not available in offline mode - check if the "Transfer owner" button is disabled

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

Same as in Test section

For example:
1. Click on the text input to bring it into focus
7. Upload an image via copy paste
8. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/7771498/5b634678-a218-4bad-86ca-e10b90abc7e5

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/7771498/34b1c875-37f7-417d-a6e9-4dc62a026cf5

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/7771498/e1ccbed9-9815-49da-90f9-2394b3a06e5e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/7771498/020252d8-1396-4935-84b3-132cf0918db3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/7771498/04db9e96-982f-4fc9-950a-9df2f6a31961

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/7771498/9b210c80-8166-4bcc-bbeb-7c8cee99c59c

</details>

burczu committed:
is selectable item property introduced

burczu committed:
showing disable icon when item is not selectable

burczu committed:
showing transfer owner button for owners opened by other admins

burczu committed:
variable names refactor

burczu committed:
constraint removed

burczu committed:
fixed constraints for an owner

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
more constraints for current user added

burczu committed:
disabling transfer owner button when offline

burczu committed:
policy ownership checks Onyx key added

burczu committed:
update workspace ownership checks method added

burczu committed:
request workspace owner change method added

burczu committed:
start change ownership flow method implemented

burczu committed:
workspace owner change check page added

burczu committed:
temporary solution to open owner check confiration page

burczu committed:
transfer balance confirmation copy handled

burczu committed:
temporary copy added

burczu committed:
showing correct owner change check page version based on error type

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
new owner payment card form page added

burczu committed:
translation key fix

burczu committed:
foundation of the add payment card form added (not finished yet))

burczu committed:
unnecessary imports removed

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
add payment card in ownership change flow design adjusted

burczu committed:
form validation added

burczu committed:
add billing card and request workspace owner change api method added

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
switched to use existing add payment card functionality

burczu committed:
comment added

burczu committed:
using correct parameter name for name on card and CardUtils

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
cleanup after last main sync

burczu committed:
copy for ownership change check page updated

burczu committed:
currency selector added to the payment card form

burczu committed:
sending currency to the API added

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
change ownership flow wrapped up

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerPaymentCardFormPage.tsx:
```diff
@@ -0,0 +1,293 @@
+    const policyID = route.params.policyID;
+    const accountID = route.params.accountID;
+
+    useEffect(
+        () => {
+            PaymentMethods.clearDebitCardFormErrorAndSubmit();
+
+            return () => {
+                PaymentMethods.clearDebitCardFormErrorAndSubmit();
+            };
+        },
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        [],
+    );
```
Why do we need this useEffect?

burczu replied:
to clear debit card data stored in Onyx while filling the form - next we open the form, it is cleared; it is done the same way we do it on the `AddDebitCardPage` page

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerPaymentCardFormPage.tsx:
```diff
@@ -0,0 +1,293 @@
+                currency: CONST.CURRENCY.USD,
+            };
+            Policy.addBillingCardAndRequestPolicyOwnerChange(policyID, cardData);
+            Navigation.dismissModal();
```
We shouldn't dismiss the modal. Instead we should let `Policy.addBillingCardAndRequestPolicyOwnerChange` decide if we should dismiss the modal , i.e. no `changeOwner` errors or navigate the user to `WorkspaceOnerChangeCheckPage`

burczu replied:
same answer as before

luacmartins reviewed (commented):
I tested all cases with the API and they seem to work well. I had to make a few changes to the `shouldTransferSingleSubscription` case and I'm also working on pushing Onyx updates to the client once the transfer is completed.

luacmartins reviewed (commented):
Here are a few bugs that I noticed:

Bug 1: Navigating back from currency page goes back to profile, instead of debit card page
 
https://github.com/Expensify/App/assets/22219519/b7b8e204-c401-4407-b754-310359aaf268

BUG 2: Navigating to the Members page opens the debit card RHP

https://github.com/Expensify/App/assets/22219519/15d75c3f-6ad2-4d32-8599-1b9916b111c5

BUG 3: Duplicate API calls

https://github.com/Expensify/App/assets/22219519/94165a27-bd72-42db-aa92-2c83673e6e6c

BUG 4: No page for failedToClearBalance error

https://github.com/Expensify/App/assets/22219519/4b4332d1-99a9-495c-af8e-cbc8e0b7ac6e

burczu commented on src/pages/workspace/members/WorkspaceOwnerChangeCheckPage.tsx:
```diff
@@ -0,0 +1,184 @@
+        if (currentPolicyOwnershipChecks) {
+            let ownershipChecks: PolicyOwnershipChangeChecks = {} as PolicyOwnershipChangeChecks;
+            if (error === CONST.POLICY.OWNERSHIP_ERRORS.AMOUNT_OWED) {
+                ownershipChecks = {
+                    ...currentPolicyOwnershipChecks,
+                    shouldClearOutstandingBalance: true,
+                };
```
ahhh... I think this is the misunderstanding that led me to do it with this `useEffect`'s etc. - I presumed that we need to store the `ownershipChecks` object between `RequestWorkspaceOwnerChage` calls... if it's not the case we, indeed, can do it in a more straightforward way - I'll change the implementation accordingly

burczu commented on src/libs/actions/Policy.ts:
```diff
@@ -974,6 +985,146 @@ function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newR
+        ...ownershipChecks,
+    };
+
+    API.write(WRITE_COMMANDS.REQUEST_WORKSPACE_OWNER_CHANGE, params, {optimisticData, successData, failureData});
```
as far as I know, using `makeRequestWithSideEffects` is highly not recommended (I agree that returning promises from actions is not best idea)... In my opinion we should keep using the `write` method here, and use `useEffect` to detect the change in the `policy` onyx object, and navigate to the next page based on the new value of the `changeOwner` error field

burczu commented on src/pages/workspace/members/WorkspaceOwnerChangeCheckPage.tsx:
```diff
@@ -0,0 +1,184 @@
+            }
+
+            Policy.updateWorkspaceOwnershipChecks(policyID, ownershipChecks);
+            Navigation.dismissModal();
```
without dismissing modal, user won't see the animation of closing and opening new modal (this is the case when we show errors based on the `changeOwner` error field) - I thought it's good from the UX point of view (user see that their action makes an effect in the UI)

burczu commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -182,6 +189,30 @@ function WorkspaceMembersPage({
+        Policy.requestWorkspaceOwnerChange(policyID);
+    }, [policyID, policyOwnershipChangeChecks]);
+
+    useEffect(() => {
```
we can't decide it in the callback assigned to the button, because we don't know this at a time - we get this information asynchronously from the API, so it needs to be done in the `useEffect` that reacts to the change of the `policy.errorFields.changeOwner` property

burczu commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -182,6 +189,30 @@ function WorkspaceMembersPage({
         getWorkspaceMembers();
     }, [isOffline, prevIsOffline, getWorkspaceMembers]);
 
+    useEffect(() => {
+        if (isEmptyObject(policyOwnershipChangeChecks?.[policyID])) {
+            return;
+        }
+
+        Policy.requestWorkspaceOwnerChange(policyID);
+    }, [policyID, policyOwnershipChangeChecks]);
```
I'm changing it - the call will be done in the button event handler

burczu commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -57,6 +58,9 @@ type WorkspaceMembersPageOnyxProps = {
     /** An object containing the accountID for every invited user email */
     invitedEmailsToAccountIDsDraft: OnyxEntry<InvitedEmailsToAccountIDs>;
+
+    /** Ownership checks */
+    policyOwnershipChangeChecks: OnyxEntry<Record<string, PolicyOwnershipChangeChecks>>;
```
getting rid of this from this page

burczu committed:
refactor the flow with assumption that ownership checks are optional

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
ownership checks display texts utility method added

burczu commented on src/pages/workspace/members/WorkspaceOwnerChangeCheckPage.tsx:
```diff
@@ -0,0 +1,184 @@
+            default:
+                return null;
+        }
+    }, [accountID, error, personalDetails, policy?.errorFields?.changeOwner, policy?.name, translate]);
```
addressed

burczu commented on src/pages/workspace/members/WorkspaceMemberDetailsPage.tsx:
```diff
@@ -52,24 +58,41 @@ function WorkspaceMemberDetailsPage({personalDetails, policyMembers, policy, rou
+
+        Policy.clearWorkspaceOwnerChangeFlow(policyID);
+        Policy.updateWorkspaceOwnershipChecks(policyID, ownershipChecks);
+        Policy.requestWorkspaceOwnerChange(policyID);
```
done

burczu committed:
ownership checks parameter made optional

burczu commented on src/libs/actions/Policy.ts:
```diff
@@ -974,6 +985,146 @@ function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newR
+                isLoading: true,
+            },
+        },
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                errorFields: null,
+                pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
+            },
+        },
```
the whole block should not be removed, because we should clear `errorFields` optimistically - other way it will keep old errors adding new ones and this will mess up the whole flow

burczu commented on src/libs/actions/Policy.ts:
```diff
@@ -974,6 +985,146 @@ function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newR
+                isLoading: false,
+            },
+        },
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                errorFields: null,
+                pendingAction: null,
+            },
+        },
```
removed

burczu commented on src/libs/actions/Policy.ts:
```diff
@@ -974,6 +985,146 @@ function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newR
+                isLoading: false,
+            },
+        },
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                pendingAction: null,
+            },
+        },
```
removed

burczu committed:
unnecessay onyx data removed

burczu commented on src/libs/actions/Policy.ts:
```diff
@@ -974,6 +985,146 @@ function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newR
+function requestWorkspaceOwnerChange(policyID: string) {
+    const ownershipChecks = policyOwnershipChecks?.[policyID] ?? {};
+
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                errorFields: null,
+                pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
+            },
+        },
+    ];
+
+    const successData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                errorFields: null,
+                pendingAction: null,
+            },
+        },
+    ];
+
+    const failureData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                pendingAction: null,
+            },
+        },
+    ];
```
changed

burczu committed:
handle isLoading for Policy added

burczu committed:
failed to clear balance error handling added (incomplete copy)

burczu committed:
currency modal closing issue fixed

burczu commented:
@luacmartins I've addressed or answered all your comments. Bugs 1 and 4 from your last comment are fixed (for Bug 4 I need texts for `failedToClearBalance` error - I've messaged you on Slack regarding this). I need some more time to investigate Bug 2 and 3.

Please also test everything one more time with the real API, because I made changes that might break the flow.

luacmartins commented on src/libs/WorkspacesSettingsUtils.ts:
```diff
@@ -211,6 +212,129 @@ function getUnitTranslationKey(unit: Unit): TranslationPaths {
     return unitTranslationKeysStrategy[unit];
 }
 
+/**
+ * @param error workspace change owner error
+ * @returns request ownership change parameters object
+ */
```
NAB don't think we need these since we have TS types

luacmartins commented on src/libs/WorkspacesSettingsUtils.ts:
```diff
@@ -211,6 +212,129 @@ function getUnitTranslationKey(unit: Unit): TranslationPaths {
+
+    const confirmationText = () => {
+        const changeOwner = policy?.errorFields?.changeOwner;
+        const subscription = changeOwner?.subscription as unknown as {ownerUserCount: number; totalUserCount: number};
+
+        switch (error) {
+            case 'noBillingCard':
+                break;
+            case CONST.POLICY.OWNERSHIP_ERRORS.AMOUNT_OWED:
+                return translate('workspace.changeOwner.amountOwedText');
+            case CONST.POLICY.OWNERSHIP_ERRORS.OWNER_OWES_AMOUNT:
+                return translate('workspace.changeOwner.ownerOwesAmountText', {
+                    email: changeOwner?.ownerOwesAmount,
+                    amount: '',
+                });
+            case CONST.POLICY.OWNERSHIP_ERRORS.SUBSCRIPTION:
+                return translate('workspace.changeOwner.subscriptionText', {
+                    usersCount: subscription?.ownerUserCount,
+                    finalCount: subscription?.totalUserCount,
+                });
+            case CONST.POLICY.OWNERSHIP_ERRORS.DUPLICATE_SUBSCRIPTION:
+                return translate('workspace.changeOwner.duplicateSubscriptionText', {
+                    email: changeOwner?.duplicateSubscription,
+                    workspaceName: policy?.name,
+                });
+            case CONST.POLICY.OWNERSHIP_ERRORS.HAS_FAILED_SETTLEMENTS:
+                return translate('workspace.changeOwner.hasFailedSettlementsText', {email: accountLogin});
+            case CONST.POLICY.OWNERSHIP_ERRORS.FAILED_TO_CLEAR_BALANCE:
+                return translate('workspace.changeOwner.failedToClearBalanceText');
+            default:
+                return null;
+        }
+    };
+
+    const title = confirmationTitle();
+    const text = confirmationText();
+    const buttonText = confirmationButtonText();
+
+    return {title, text, buttonText};
+}
```
Hmm I think what I had in mind for this was something like:
```suggestion
function getOwnershipChecksDisplayText(
    error: ValueOf<typeof CONST.POLICY.OWNERSHIP_ERRORS>,
    translate: <TKey extends TranslationPaths>(phraseKey: TKey, ...phraseParameters: PhraseParameters<Phrase<TKey>>) => string,
    policy: OnyxEntry<Policy>,
    accountLogin: string | undefined,
) {
    let title;
    let text;
    let buttonText;

    const changeOwner = policy?.errorFields?.changeOwner;
    const subscription = changeOwner?.subscription as unknown as {ownerUserCount: number; totalUserCount: number};

    switch (error) {
        case CONST.POLICY.OWNERSHIP_ERRORS.AMOUNT_OWED:
            title = translate('workspace.changeOwner.amountOwedTitle');
            text = translate('workspace.changeOwner.amountOwedText');
            buttonText = translate('workspace.changeOwner.amountOwedButtonText');
            break;
        case CONST.POLICY.OWNERSHIP_ERRORS.OWNER_OWES_AMOUNT:
            title = translate('workspace.changeOwner.ownerOwesAmountTitle');
            text = translate('workspace.changeOwner.ownerOwesAmountText', {email: changeOwner?.ownerOwesAmount, amount: ''});
            buttonText = translate('workspace.changeOwner.ownerOwesAmountButtonText');
            break;
        case CONST.POLICY.OWNERSHIP_ERRORS.SUBSCRIPTION:
            title = translate('workspace.changeOwner.subscriptionTitle');
            text = translate('workspace.changeOwner.subscriptionText', {usersCount: subscription?.ownerUserCount, finalCount: subscription?.totalUserCount});
            buttonText = translate('workspace.changeOwner.subscriptionButtonText');
            break;
        case CONST.POLICY.OWNERSHIP_ERRORS.DUPLICATE_SUBSCRIPTION:
            title = translate('workspace.changeOwner.duplicateSubscriptionTitle');
            text = translate('workspace.changeOwner.duplicateSubscriptionText', {email: changeOwner?.duplicateSubscription, workspaceName: policy?.name});
            buttonText = translate('workspace.changeOwner.duplicateSubscriptionButtonText');
            break;
        case CONST.POLICY.OWNERSHIP_ERRORS.HAS_FAILED_SETTLEMENTS:
            title = translate('workspace.changeOwner.hasFailedSettlementsTitle');
            text = translate('workspace.changeOwner.hasFailedSettlementsText', {email: accountLogin});
            buttonText = translate('workspace.changeOwner.hasFailedSettlementsButtonText');
            break;
        case CONST.POLICY.OWNERSHIP_ERRORS.FAILED_TO_CLEAR_BALANCE:
            title = translate('workspace.changeOwner.failedToClearBalanceTitle');
            text = translate('workspace.changeOwner.failedToClearBalanceText');
            buttonText = translate('workspace.changeOwner.failedToClearBalanceButtonText');
            break;
        default:
            title = '';
            buttonText = '';
            text = '';
            break;
    }

    return {title, text, buttonText};
}
```

burczu replied:
changed

luacmartins commented on src/pages/workspace/members/WorkspaceMemberDetailsPage.tsx:
```diff
@@ -52,24 +59,49 @@ function WorkspaceMemberDetailsPage({personalDetails, policyMembers, policy, rou
+            return;
+        }
+
+        const keys = Object.keys(policy.errorFields.changeOwner);
```
```suggestion
        const changeOwnerErrors = Object.keys(policy.errorFields.changeOwner);
```

burczu replied:
changed

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerChangeCheckPage.tsx:
```diff
@@ -0,0 +1,110 @@
+            return;
+        }
+
+        const keys = Object.keys(policy.errorFields.changeOwner);
```
```suggestion
        const changeOwnerErrors = Object.keys(policy.errorFields.changeOwner);
```

burczu replied:
changed

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerChangeCheckPage.tsx:
```diff
@@ -0,0 +1,110 @@
+    const error = route.params.error;
+
+    useEffect(() => {
+        if (!policy) {
```
The logic in this useEffect seems to be the same as the one in `WorkspaceMemberDetailsPage`. Can we extract it into a method and call that method in both places?

burczu replied:
extracted

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerChangeCheckPage.tsx:
```diff
@@ -0,0 +1,110 @@
+        const ownershipChecks = WorkspaceSettingsUtils.getOwnershipChecks(error);
+
+        PolicyActions.requestWorkspaceOwnerChange(policyID, ownershipChecks);
+        Navigation.dismissModal();
```
I don't think we should dismiss this modal. We should keep it visible and show a loading indicator while the request is pending

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerPaymentCardFormPage.tsx:
```diff
@@ -0,0 +1,314 @@
+    );
+
+    useEffect(() => {
+        if (!policy) {
```
Same for the logic here

burczu replied:
done

luacmartins commented on src/libs/WorkspacesSettingsUtils.ts:
```diff
@@ -211,6 +212,129 @@ function getUnitTranslationKey(unit: Unit): TranslationPaths {
+ * @param error workspace change owner error
+ * @returns request ownership change parameters object
+ */
+function getOwnershipChecks(error: ValueOf<typeof CONST.POLICY.OWNERSHIP_ERRORS>) {
```
I don't think we need this method. I think we should instead retrieve it directly from the policy in actions/Policy.ts. I left a comment about it below

burczu replied:
addressed

luacmartins commented on src/libs/actions/Policy.ts:
```diff
@@ -982,6 +985,117 @@ function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newR
 }
 
+function requestWorkspaceOwnerChange(policyID: string, ownershipChecks?: PolicyOwnershipChangeChecks) {
+    const optimisticData: OnyxUpdate[] = [
```
I think we should retrieve `ownershipChecks` from the `POLICY_OWNERSHIP_CHANGE_CHECKS` Onyx key as we described in the doc, but I can't see that implemented here

luacmartins replied:
We should always be passing all previously accepted params, i.e. `shouldTransferAmountOwed`, `shouldTransferSubscription`, etc otherwise the API might return an error that was already accepted by the user

burczu replied:
alright, now it's all clear, I think - I've brought back storing `ownershipChecks` and handle it only in the `requestWorkspaceOwnerChange`

luacmartins reviewed (changes requested):
There are a few bugs that I've noticed as well:

1. Cursor on currency input

https://github.com/Expensify/App/assets/22219519/c72db58c-cb09-48b0-8b0f-40fbde2260f2

2. We navigate away from the change owner checks modal and then reopen it. This happens for all errors

https://github.com/Expensify/App/assets/22219519/c4d5b697-55fc-4a0e-9fba-0a6cefd6b655

3. It's possible to get stuck in a loop because we don't pass previously accepted terms, e.g. `shouldTransferSubscription` and `shouldTransferSingleSubscription`

https://github.com/Expensify/App/assets/22219519/776a6015-1eb3-4bbf-a7fa-634270d03a7d

burczu committed:
getting ownership texts method fixed

burczu committed:
key variable name amended

burczu committed:
common useEffect method extracted as utility function

burczu committed:
loading indicator introduced

burczu committed:
storing ownership checks in ONYX brought back and re-designed

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

luacmartins commented on src/libs/WorkspacesSettingsUtils.ts:
```diff
@@ -212,6 +215,92 @@ function getUnitTranslationKey(unit: Unit): TranslationPaths {
+            buttonText = translate('workspace.changeOwner.failedToClearBalanceButtonText');
+            break;
+        default:
+            title = '';
+            buttonText = '';
```
```suggestion
            title = '';
            text = '';
            buttonText = '';
```

burczu replied:
fixed

luacmartins reviewed (changes requested):
Code is looking good! There are still a couple of things we can improve:
 
1. We can get into a loop. Take the `hasFailedSettlement` as an example. There's no in-product user action, so pressing `Got it` should NOT make another API call, clear the errors and navigate the user back. Otherwise they get navigated back to the RHP when then go back.

https://github.com/Expensify/App/assets/22219519/5c1c64d7-4681-4930-bf7d-20ea9971a2e9

2. I think we should introduce a generic error screen for when we don't know what's wrong. In the example below, we get an unknown error from the server, but we don't display anything to the user. I think we should display a generic error and either ask them to try again later or contact Concierge for assistance.

https://github.com/Expensify/App/assets/22219519/58e85aae-676d-47f8-803c-f96135a5ba3f

3. I think we should add a success screen. Right now there's no feedback to the user that they are now the owner of the workspace other than the `Transfer owner` button changing to `Remove from Workspace`. Additionally, I think pressing the `Change owner` button should always navigate the user to the `WorkspaceOwnerChangeChecks` screen with a loading indicator so there's feedback to the user that we're processing their request and if there are no errors display the success page, otherwise display the existing error screens.

https://github.com/Expensify/App/assets/22219519/7bed3034-6530-4c35-abe2-ba587fed5d9d

Tagging @Expensify/design for thoughts on 2 and 3.

trjExpensify commented:
Ah yeah, interesting one. 

>I think we should add a success screen. Right now there's no feedback to the user that they are now the owner of the workspace other than the Transfer owner button changing to Remove from Workspace.

Yeah, I agree. The growls would have been nice here, but that pattern has gone away. So how about something like this for the success screen? (Design to approve! =D)

![image](https://github.com/Expensify/App/assets/16232057/c0ed0e85-e34b-455b-ad56-02f759f6e74d)

>I think we should introduce a generic error screen for when we don't know what's wrong. In the example below, we get an unknown error from the server, but we don't display anything to the user. I think we should display a generic error and either ask them to try again later or contact Concierge for assistance.

We could do the same thing here with a screen for if it fails, something like the below with the hyperlink on Concierge going to the Conci DM:

![image](https://github.com/Expensify/App/assets/16232057/5c1eb29e-8403-4b86-b354-6eaa7647a5c2)

Though, I guess another option is adding an in-line dismissible error beneath the `Transfer owner` button when they return to that screen or something?

luacmartins commented:
Yea, I like those. The issue with the inline error is that we'd still not show any feedback while the request is pending, i.e. that button has no loading state. I like the screens shared above!

dannymcclain commented:
I like those too.

On the `Oops! Not so fast...` one, does clicking "Got it" just dismiss the screen and essentially abandon the transfer owner flow?

Also on that same screen, I think we should make `reach out to Concierge` the link instead of just "Concierge" (just to tie the action into the interactive element).

Lastly, would it be worth making the money guy not look SO happy? Probably not a big deal, but it is kinda funny. 😂 
![CleanShot 2024-03-21 at 14 03 42@2x](https://github.com/Expensify/App/assets/7516624/2766db6f-5b4d-4f3e-9b6a-76e87249b66f)

trjExpensify commented:
>On the Oops! Not so fast... one, does clicking "Got it" just dismiss the screen and essentially abandon the transfer owner flow?

Yep! They'll return to here: 

![image](https://github.com/Expensify/App/assets/16232057/d01fd6eb-b573-4cad-aac1-4c1991e34184)

>Also on that same screen, I think we should make reach out to Concierge the link instead of just "Concierge" (just to tie the action into the interactive element).

Perfect!

>Lastly, would it be worth making the money guy not look SO happy? Probably not a big deal, but it is kinda funny. 😂

Love that!! I just really liked using him because he had his hand raised like "sthap! ✋" 

![image](https://github.com/Expensify/App/assets/16232057/0d8a1c9c-d03d-40e5-9804-595df8245c88)

luacmartins commented:
Gonna open this up for C+ to review the code to get a head start. I think we can re-review once we add those additional screens.

dubielzyk-expensify commented:
Love that detail @dannymcclain !

Can we also center the description text? Unless this is a new thing we're doing now. 
 
![CleanShot 2024-03-22 at 08 38 51@2x](https://github.com/Expensify/App/assets/145416827/a99eace7-e79e-455d-b4c0-440eb53fbeab)

trjExpensify commented:
About 99.9% sure that's not something we're doing. Can you design proof these bad boys? https://www.figma.com/file/O2SBnfaTdx4OBpFQMf9JIt?node-id=1325:39905&mode=design#744265015

burczu committed:
hidding carret of the currency combo box

burczu committed:
missing variable setting added

burczu committed:
unexpected loop for failed transfer fixed

burczu committed:
change owner -> transfer owner wording amended

burczu commented:
@trjExpensify @dannymcclain @dubielzyk-expensify Question about the success screen. Should I use this new static `fireworks` from the desing or use already existing fireworks animation we have in the App? It would look like below with the animation:

https://github.com/Expensify/App/assets/7771498/a3fce2cf-30bc-4c41-8c46-1ca69fe866ba

burczu committed:
ownership change success page added

trjExpensify commented:
Love that, use it I say!

burczu committed:
ownership change flow generic error page added

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu commented:
@luacmartins I think I've fixed previously raised errors like cursor in currency field or loop issue. I've also implemented two new pages for error and success. Unfortunately, introducing them caused some unexpected loops again (you'll probably notice it while testing) and I need some more time to fix it...

luacmartins commented:
Thank you! I'll test it again today :)

dannymcclain commented:
@trjExpensify @dubielzyk-expensify I think we _do_ center the description? Here are some examples:

LHN empty state:
![CleanShot 2024-03-22 at 08 20 46@2x](https://github.com/Expensify/App/assets/7516624/d2f21349-39b1-423c-b12a-66047d2ca741)

Desktop scan flow:
![CleanShot 2024-03-22 at 08 20 32@2x](https://github.com/Expensify/App/assets/7516624/b93d176d-ba00-4197-a862-654722a7d811)

Forced update:
![CleanShot 2024-03-22 at 08 22 10@2x](https://github.com/Expensify/App/assets/7516624/03157ac6-3e36-4a6d-ae26-8e4b588d96b2)

There are probably others but these came to the top of my mind.

trjExpensify commented:
Ah, I think we're all saying they should be centered, but my example in Figma was off or something. :)

dannymcclain commented:
🙈 I misunderstood the comment thread—my bad!

mollfpr commented:
So far the file changes look good to me!

luacmartins commented on src/languages/en.ts:
```diff
@@ -2086,6 +2088,45 @@ export default {
+            addPaymentCardSecurity: 'security',
+            amountOwedTitle: 'Outstanding balance',
+            amountOwedButtonText: 'OK',
+            amountOwedText: 'This account has an outstanding balance from a previous month.\n\nDo you want to clear balance and take over billing of this workspace?',
```
I think the [copy](https://expensify.slack.com/archives/C21FRDWCV/p1710521136442169?thread_ts=1710435317.182139&cid=C21FRDWCV) included the amount
```suggestion
            amountOwedText: ({amount}) => `This account has an outstanding balance of ${amount} from a previous month.\n\nDo you want to clear balance and take over billing of this workspace?`,
```

burczu replied:
addressed

luacmartins reviewed (changes requested):
Looking good. As you mentioned, we still have a few issues:

1. Users are not navigated to the checks page with a loading indicator once they press `Transfer owner`

https://github.com/Expensify/App/assets/22219519/00caabda-0bf3-47cd-8e70-4a10a82a6376

2. Loop on `failedToClearBalance`

https://github.com/Expensify/App/assets/22219519/78345b8a-d76e-43bd-82c8-723208e2ebb5

3. Loop on `hasFailedSettlements

https://github.com/Expensify/App/assets/22219519/29c6f398-2b08-48ac-9a44-4caece12958f

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

burczu committed:
is successful flag introduced to prevent loops

burczu committed:
showing owner change check modal right away

burczu committed:
owner owes amount error handling adjusted

burczu commented on src/libs/WorkspacesSettingsUtils.ts:
```diff
@@ -212,6 +215,100 @@ function getUnitTranslationKey(unit: Unit): TranslationPaths {
+            title = translate('workspace.changeOwner.ownerOwesAmountTitle');
+            text = translate('workspace.changeOwner.ownerOwesAmountText', {
+                email: changeOwner?.ownerOwesAmount,
+                amount: '',
```
addressed

burczu committed:
missing useCallback deps added

shawnborton commented:
Where is this mockup coming from?
![CleanShot 2024-03-25 at 10 11 42@2x](https://github.com/Expensify/App/assets/2319350/0b37e247-4aa5-40a5-a648-21ee9d0f36bb)

It feels like a bit of an outlier in terms of style.

shawnborton commented:
The "Woohoo" area doesn't quite feel vertically centered in the space below the top page header area and the bottom green button. Can you check that out? 
![CleanShot 2024-03-25 at 10 15 13@2x](https://github.com/Expensify/App/assets/2319350/a51bd86a-7110-4207-a076-4b1059c754a1)

trjExpensify commented:
> I wonder if we should put an explainer sentence under the headline here, so it's just a little more clear why we are asking you to enter payment details in order to make the transfer happen?

Whatcha' mean? Like something more than "Enter your payment card to transfer ownership" title, or is that missing in the build?

luacmartins commented on src/ROUTES.ts:
```diff
@@ -612,6 +612,23 @@ const ROUTES = {
         getRoute: (policyID: string, accountID: number, backTo?: string) => getUrlWithBackToParam(`settings/workspaces/${policyID}/members/${accountID}/role-selection`, backTo),
     },
+    WORKSPACE_OWNER_PAYMENT_CARD_FORM: {
+        route: 'settings/workspaces/:policyID/add-payment-card/:accountID',
```
I'm thinking that we should just make this route be part of the change checks, e.g.  `settings/workspaces/:policyID/change-owner/:accountID/:error` and in this case error would be `add-payment-card`. This would avoid the transition that happens from loading indicator to form page.

https://github.com/Expensify/App/assets/22219519/7eec6076-cf9a-41d9-9566-5abb31e47ab4

burczu replied:
yeah, makes sense - unified these two pages into one

luacmartins commented on src/pages/workspace/members/WorkspaceOwnerChangeErrorPage.tsx:
```diff
@@ -0,0 +1,77 @@
+                        />
+                        <Text style={[styles.textHeadline, styles.textAlignCenter, styles.mv2]}>Oops! Not so fast...</Text>
+                        <Text style={[styles.textAlignCenter, styles.textSupporting]}>
+                            There was a problem transferring ownership of this workspace. Try again, or <TextLink href={`mailto:${CONST.EMAIL.CONCIERGE}`}>reach out to Concierge</TextLink>{' '}
```
We need to translate this

luacmartins reviewed (changes requested):
Looking better. I left a couple of comments. In addition, we have the following issues:

1. Unknown error page is not shown. I couldn't get to the unknown error page even when I got a `405 Resource in incorrect state` error

https://github.com/Expensify/App/assets/22219519/0f7fdc13-95d9-4a23-8f2f-8cd0746c10c4

2. Stale error shows momentarily when navigating and before loading spinner

https://github.com/Expensify/App/assets/22219519/dbd412be-582b-42dc-9d6d-1926d307f7f8

luacmartins commented:
It seems like the TS error might be related to cache. I see that we disable the default export error [here](https://github.com/Expensify/App/blob/49cc62419bee1cb1edb9d02c488064dd4c465439/workflow_tests/mocks/failureNotifierMocks.ts#L15).

```
Error: workflow_tests/failureNotifier.test.ts(6,8): error TS1192: Module '"/home/runner/work/App/App/workflow_tests/mocks/failureNotifierMocks"' has no default export.
```

burczu committed:
unexpected errors handling fixed

burczu committed:
using currency utils to format amount

burczu commented on src/libs/WorkspacesSettingsUtils.ts:
```diff
@@ -244,8 +245,8 @@ function getOwnershipChecksDisplayText(
-                email: changeOwner?.ownerOwesAmount,
-                amount: '',
+                email: ownerOwesAmount?.ownerEmail,
+                amount: `${ownerOwesAmount?.amount} ${ownerOwesAmount?.currency}`,
```
done

burczu committed:
missing translations added

burczu committed:
refactor: payment form and check page unified

burczu committed:
handling transition end to prevent showing stale error on first transition

burczu commented:
@luacmartins I've just addressed your last comments including a refactor to handle payment card form inside the common change owner check page. Regarding bugs, you've raised:

> 1. Unknown error page is not shown. I couldn't get to the unknown error page even when I got a 405 Resource in incorrect state error

This was because I had to work with mocked API, and my assumption was that I'll get an `errorFields` object with some unexpected error messages - based on the example you've shown in your comment, it looks like in this case the `errorFields` object is cleared, so I've change the implementation accordingly and it should (hopefully) work now as expected. Please test it one more time.

> 2. Stale error shows momentarily when navigating and before loading spinner

I think I've managed to fix this for the case when we open the modal for the first time (like you've shown in the screen capture video). But there is still a glitch when we switch between errors in the opened modal - in some cases after the loader disappears, we see the previous texts for milliseconds. I didn't find out yet how to fix it, but maybe we could try to address it in a follow up issue?

burczu commented:
@shawnborton regarding your comments:

> The "Woohoo" area doesn't quite feel vertically centered in the space below the top page header area and the bottom green button. Can you check that out?

It is vertically centered:

![Screenshot 2024-03-26 at 11 02 22](https://github.com/Expensify/App/assets/7771498/b4c2b600-35c4-4780-913a-438721e32bb1)

The feeling that there is something wrong with the alignment comes from the fact that there is a lot of empty space in the animation/image. I've used existing component for this that we use, e.g., in the enable payment process or in the wallet's transfer balance page.

> For these screens:
> [CleanShot 2024-03-25 at 10 14 08@2x](https://private-user-images.githubusercontent.com/2319350/316529501-ab517afd-5f50-47a0-a111-3e58ca38aa97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NDQ4ODgsIm5iZiI6MTcxMTQ0NDU4OCwicGF0aCI6Ii8yMzE5MzUwLzMxNjUyOTUwMS1hYjUxN2FmZC01ZjUwLTQ3YTAtYTExMS0zZTU4Y2EzOGFhOTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjZUMDkxNjI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWUyMGFiNDJkOTRjZWQ4MzdkMmEzNDVmNmU3NWFiNjJjN2Y4Nzc2ZDI0YWUyMjY2NDU2ZDllODZkOGYyYzk3YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.D7PilvSKilW2oqw8PYo4ZDxjE6HzpH6fyqICY-7ij4M)

> Can we tighten up the gap between headline and text? Maybe try 8px gap below the headline?

Do you mean the gap between the "Transfer owner" and "Outstanding balance" or between "Outstanding balance" and "The account owning the workspace..." text?

shawnborton commented:
> The feeling that there is something wrong with the alignment comes from the fact that there is a lot of empty space in the animation/image. I've used existing component for this that we use, e.g., in the enable payment process or in the wallet's transfer balance page.

Sounds good, thanks for confirming. 

> or between "Outstanding balance" and "The account owning the workspace..." text?

This, the headline and the text immediately below it.

burczu commented:
@shawnborton Thanks for the answer. I've just changed the gap to 8px, as requested - it will look like below:

![Screenshot 2024-03-26 at 14 34 35](https://github.com/Expensify/App/assets/7771498/e670a081-63d6-4ecb-9aeb-bf46f180c4fe)
![Screenshot 2024-03-26 at 14 36 53](https://github.com/Expensify/App/assets/7771498/f0724b7a-b334-480e-919b-7b7ef75958cd)

burczu committed:
styles adjustment

burczu committed:
Merge branch 'main' into feature/35716-changing-owner-flow

shawnborton commented:
> Thanks for the answer. I've just changed the gap to 8px, as requested - it will look like below:

Looks good, thanks!

luacmartins reviewed (approved):
Code LGTM. I'm still seeing a few small things that we need to address, but those can be done in a follow up:

1. Disable `Transfer owner` button if user is offline since we can only go through this flow while online.

2. Wrong error flashes momentarily

https://github.com/Expensify/App/assets/22219519/5aa30edd-7614-4249-9751-5ecfaf636d73

3. We're showing an error screen on success after adding a debit card

https://github.com/Expensify/App/assets/22219519/99212ff8-ddf7-4a47-a45a-3ee3b2e95ebd

4. We're showing three decimals for USD currency

![three_decimals](https://github.com/Expensify/App/assets/22219519/43705f7f-0881-4c27-8d84-cff5f6c5023a)

luacmartins commented:
I'll do the checklist on this one since it's hard for C+ to test it because we need the API in certain states to trigger errors.

luacmartins commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

burczu commented on src/pages/workspace/members/WorkspaceOwnerPaymentCardCurrencyModal.tsx:
```diff
@@ -0,0 +1,85 @@
+                    onBackButtonPress={onClose}
+                />
+                <SelectionList
+                    shouldDelayFocus
```
ach, noo - must have been added when I was trying to fix one of the issues and forgotten to remove it. I'll remove it in while working on the [follow up issue](https://github.com/Expensify/App/issues/39005). Thanks for pointing this out.

iwiznia commented on src/languages/es.ts:
```diff
@@ -2126,6 +2128,50 @@ export default {
+            hasFailedSettlementsButtonText: 'Entiendo',
+            hasFailedSettlementsText: ({email}) =>
+                `No puede hacerse cargo de la facturación porque ${email} tiene una liquidación vencida de la tarjeta Expensify. Avíseles que se comuniquen con concierge@expensify.com para resolver el problema. Luego, podrá hacerse cargo de la facturación de este espacio de trabajo.`,
+            failedToClearBalanceTitle: 'Failed to clear balance',
+            failedToClearBalanceButtonText: 'OK',
+            failedToClearBalanceText: 'We were unable to clear the balance. Please try again later.',
```
How come this is in english? This is against the checklists @burczu @luacmartins

luacmartins replied:
Damn, I thought we had updated all of them, but clearly we missed some. I'll put up a PR to fix it. Thanks for raising this.

jayeshmangwani commented on src/pages/workspace/members/WorkspaceOwnerChangeWrapperPage.tsx:
```diff
@@ -0,0 +1,100 @@
+        const changeOwnerErrors = Object.keys(policy?.errorFields?.changeOwner ?? {});
+
+        if (changeOwnerErrors && changeOwnerErrors.length > 0 && changeOwnerErrors[0] !== CONST.POLICY.OWNERSHIP_ERRORS.NO_BILLING_CARD) {
+            Navigation.navigate(ROUTES.WORKSPACE_OWNER_CHANGE_CHECK.getRoute(policyID, accountID, changeOwnerErrors[0] as ValueOf<typeof CONST.POLICY.OWNERSHIP_ERRORS>));
```
coming from to this issue https://github.com/Expensify/App/issues/57245, when we encounter an `changeOwnerErrors` value, we navigate to `WORKSPACE_OWNER_CHANGE_CHECK`. However, this leads to the same page being added to the navigation stack since we are already on `WORKSPACE_OWNER_CHANGE_CHECK`. As a result, pressing the back button on the success transfer page takes us to a blank page because the same screen exists twice with different parameters and to reach the member details page, we then have to press back twice. Instead, we can simply update the params of the current page if we receive a different error value from the backend. More context can be found here: https://github.com/Expensify/App/issues/57245#issuecomment-2676151355
